singleton-service/src/service/auth.rs

```diff
@@ -0,0 +1,86 @@
+use log::debug;
+use threescale::structs::AppIdentifier;
+use threescalers::{
+    api_call::{ApiCall, Kind},
+    application::Application,
+    credentials::*,
+    extensions::{self},
+    http::Request,
+    service::Service,
+    transaction::Transaction,
+};
+
+/// Proxy level representation for authorization request.
+pub struct Auth {
+    pub service_id: String,
+    pub service_token: String,
+    pub app_id: AppIdentifier,
+}
+
+impl Auth {
+    pub fn service_id(&self) -> &str {
+        self.service_id.as_str()
+    }
+
+    pub fn service_token(&self) -> &str {
+        self.service_token.as_str()
+    }
+
+    pub fn app_id(&self) -> &AppIdentifier {
+        &self.app_id
+    }
+}
+
+/// Create a vector of Auth objects for a service. Take service_key(service_id + service_token)
+/// and apps_keys of type Vec<AppIdentifier> as arguments to the function and returns Vec<Auth>.
+pub fn auth_apps(service_key: String, app_keys: Vec<AppIdentifier>) -> Vec<Auth> {
+    let keys = service_key.split('_').collect::<Vec<_>>();
+    app_keys
+        .iter()
+        .map(|app| auth(keys[0].to_string(), keys[1].to_string(), app.clone()))
+        .collect::<Vec<_>>()
+}
+
+/// Create a Auth object for an application. Take service_id, service_token and app_id of type
+/// AppIdentifier and returns an Auth object.
+pub fn auth(service_id: String, service_token: String, app_id: AppIdentifier) -> Auth {
+    Auth {
+        service_id,
+        service_token,
+        app_id,
+    }
+}
+
+/// Create a Request of type Authorize. Take an object of type Auth as argument to the function
+/// and returns Result<Request, anyhow::Error>.
+pub fn build_auth_request(auth: &Auth) -> Result<Request, anyhow::Error> {
+    let creds = Credentials::ServiceToken(ServiceToken::from(auth.service_token()));
+    let svc = Service::new(auth.service_id(), creds);
+    let app;
+    match auth.app_id() {
+        AppIdentifier::UserKey(_user_key) => {
+            debug!("AppIdentifier UserKey: {}", auth.app_id().as_string());
+            app = Application::from_user_key(auth.app_id().as_string())
+        }
+        AppIdentifier::AppId(_app_id, None) => {
+            debug!("AppIdentifier AppId : {}", auth.app_id().as_string());
+            app = Application::from_app_id(auth.app_id().as_string())
+        }
+        AppIdentifier::AppId(_app_id, _app_key) => {
+            let app_key = auth.app_id().as_string();
+            debug!("AppIdentifier AppId+AppKey : {}", app_key);
+            let keys = app_key.split('_').collect::<Vec<_>>();
+            app = Application::from_app_id_and_key(keys[0], keys[1])
+        }
+    }
```

Comment:
Can you please shed some light on the following concerns:
- You've pattern-matched `user_key` but not used it instead provided app_id in place of it?
- Even in the second pattern? you have required fields but still reforming them using `as_string()`?
- Same goes for 3rd pattern.
Maybe I am getting this wrong?

Reply:
sorry this was a temp thing i was trying and forgot to fix it before the PR. fixed